Marlin/planner.cpp

```diff
@@ -1301,127 +1301,162 @@ void Planner::_buffer_steps(const int32_t (&target)[XYZE]
-       Solve for max velocity based on max acceleration about the radius of the circle, defined
-       indirectly by junction deviation.
+      // Unit vector of previous path line segment
+      static float previous_unit_vec[
+        #if ENABLED(JUNCTION_DEVIATION_INCLUDE_E)
+          XYZE
+        #else
+          XYZ
+        #endif
+      ];
 
-       This may be also viewed as path width or max_jerk in the previous grbl version. This approach
-       does not actually deviate from path, but used as a robust way to compute cornering speeds, as
-       it takes into account the nonlinearities of both the junction angle and junction velocity.
-     */
+      float unit_vec[] = {
+        delta_mm[A_AXIS] * inverse_millimeters,
+        delta_mm[B_AXIS] * inverse_millimeters,
+        delta_mm[C_AXIS] * inverse_millimeters
+        #if ENABLED(JUNCTION_DEVIATION_INCLUDE_E)
+          , delta_mm[E_AXIS] * inverse_millimeters
+        #endif
+      };
 
     vmax_junction = MINIMUM_PLANNER_SPEED; // Set default max junction speed
 
     // Skip first block or when previous_nominal_speed is used as a flag for homing and offset cycles.
     if (moves_queued && !UNEAR_ZERO(previous_nominal_speed)) {
       // Compute cosine of angle between previous and current path. (prev_unit_vec is negative)
       // NOTE: Max junction velocity is computed without sin() or acos() by trig half angle identity.
-      const float cos_theta = - previous_unit_vec[X_AXIS] * unit_vec[X_AXIS]
-                              - previous_unit_vec[Y_AXIS] * unit_vec[Y_AXIS]
-                              - previous_unit_vec[Z_AXIS] * unit_vec[Z_AXIS];
-      // Skip and use default max junction speed for 0 degree acute junction.
-      if (cos_theta < 0.95) {
-        vmax_junction = min(previous_nominal_speed, block->nominal_speed);
-        // Skip and avoid divide by zero for straight junctions at 180 degrees. Limit to min() of nominal speeds.
-        if (cos_theta > -0.95) {
-          // Compute maximum junction velocity based on maximum acceleration and junction deviation
-          float sin_theta_d2 = SQRT(0.5 * (1.0 - cos_theta)); // Trig half angle identity. Always positive.
-          NOMORE(vmax_junction, SQRT(block->acceleration * junction_deviation * sin_theta_d2 / (1.0 - sin_theta_d2)));
-        }
+      float junction_cos_theta = -previous_unit_vec[X_AXIS] * unit_vec[X_AXIS]
+                                 -previous_unit_vec[Y_AXIS] * unit_vec[Y_AXIS]
+                                 -previous_unit_vec[Z_AXIS] * unit_vec[Z_AXIS]
+                                  #if ENABLED(JUNCTION_DEVIATION_INCLUDE_E)
+                                    -previous_unit_vec[E_AXIS] * unit_vec[E_AXIS]
+                                  #endif
+                                ;
+
+      // NOTE: Computed without any expensive trig, sin() or acos(), by trig half angle identity of cos(theta).
+      if (junction_cos_theta > 0.999999) {
```

Comment:
We have already set vmax_junction to MINIMUM_PLANNER_SPEED above, we could invert the if and skip this block?

```
if (junction_cos_theta < 0.999999) {
   junction_cos_theta = max(junction_cos_theta, -0.999999); // Check for numerical round-off to avoid divide by zero.
   etc...
```

Reply:
Patched.

Reply:
It was on my list... 😄